pkg/quarterdeck/auth.go

```diff
@@ -637,6 +639,126 @@ func (s *Server) Refresh(c *gin.Context) {
+// another organization. The user must submit the orgID of the organization they wish
+// to switch to and the user must belong to that organization otherwise an error is
+// returned.
+//
+// NOTE: this endpoint cannot be used with api keys because api keys are only ever
+// issued to one organization (and in fact, one project inside of one organization).
+// Only human users can belong to multiple organizations.
+func (s *Server) Switch(c *gin.Context) {
+	var (
+		err error
+		in  *api.SwitchRequest
+		out *api.LoginReply
+	)
+
+	if err = c.BindJSON(&in); err != nil {
+		sentry.Warn(c).Err(err).Msg("could not parse refresh request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse(api.ErrUnparsable))
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "unparseable").Inc()
+		return
+	}
+
+	// Ensure the orgID is included in the request
+	if ulids.IsZero(in.OrgID) {
+		log.Debug().Msg("missing orgID in switch request")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("missing organization id"))
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "missing org_id").Inc()
+		return
+	}
+
+	// Parse the claims from the access token in the request
+	var claims *tokens.Claims
+	if claims, err = middleware.GetClaims(c); err != nil {
+		sentry.Error(c).Err(err).Msg("could not get user claims from authenticated request")
+		c.JSON(http.StatusUnauthorized, api.ErrorResponse(api.ErrInvalidUserClaims))
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "no access token").Inc()
+		return
+	}
+
+	// If the orgID in the claims is the same as the requested orgID return an error
+	// (the user must use the refresh endpoint to get claims for the same orgID)
+	if orgID := claims.ParseOrgID(); orgID.Compare(in.OrgID) == 0 {
+		sentry.Warn(c).Msg("user attempting to switch to the same organization")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("cannot switch into the organization you are currently logged into"))
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "switch to same org").Inc()
+		return
+	}
+
+	// Verify that the user belongs to the requested organization.
+	var orgUser *models.OrganizationUser
+	if orgUser, err = models.GetOrgUser(c.Request.Context(), claims.Subject, in.OrgID); err != nil {
+		if errors.Is(err, db.ErrNotFound) {
+			sentry.Warn(c).Str("userID", claims.Subject).Str("orgID", in.OrgID.String()).Msg("user attempt to switch into organization they do not belong to")
+			c.JSON(http.StatusForbidden, api.ErrorResponse("invalid credentials"))
+			metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "invalid credentials").Inc()
+			return
+		}
+
+		sentry.Error(c).Err(err).Msg("could not retrieve organization user from database")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not process switch request"))
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "unhandled error").Inc()
+		return
+	}
```

Comment:
Two methods here:

1. retrieve org user from database (As done here)
2. retrieve user and call `SwitchOrganization`

I debated on which one to use; `SwitchOrganization` is used in Login so now I'm leaning towards that. Any thoughts?

Reply:
Perhaps we should use the `SwitchOrganization` since it abstracts a bit more? Then this method doesn't have to know what an `OrganizationUser` is.

Reply:
Ok sounds good; I'll change the code to use `SwitchOrganization`. Will try to do it when I make it to Miami.